Add long text reference support for links, phone and email

melloskitten opened the pull request:
This PR makes links, phones and email references in the app information interactable. This was an accessability request mentioned in this JIRA ticket [here](https://jira.itc.sap.com/browse/EXPOSUREAPP-1595).

Here's a screenshot from the Terms of Use that shows what I mean:

![image](https://user-images.githubusercontent.com/15066374/88385539-d46ae400-cdae-11ea-8b48-f5976d1b35c8.png)

AdnanZiSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationDetailViewController.swift:
```diff
@@ -57,7 +57,7 @@ extension AppInformationDetailViewController {
 extension AppInformationDetailViewController: UITextViewDelegate {
 	func textView(_ textView: UITextView, shouldInteractWith url: URL, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
-		WebPageHelper.openSafari(withUrl: url, from: self)
+		UIApplication.shared.open(url, options: [:], completionHandler: nil)
```
What was the reason to change this? Links were opened with the old code too.

johannesrohwer replied:
i assume the other implementation did not work for telephone numbers?

EisvogelSAP replied:
This would need extra handling in this case. Hyperlinks should not jump out of the app, should they?

AdnanZiSAP replied:
That was also my concern. In other places (e.g., Onboarding Privacy), links are opened inside the app. With this, links are opened outside of the app. So the behaviour is changed.

melloskitten replied:
Will add extra handling in the existing method, then no crashes should occur (this comment kinda relates to all the other comments you guys have done 👍 )

AdnanZiSAP replied:
This was not addressed in the new code. Links are still jumping outside of the app. 

IMO this needs to be aligned with the PO, and either:

1. All links in the app are opened inside the app, or
2. All links in the app are opened outside the app.

Personally, in this point of time, I wouldn't align anything, I would just implement 1., as it's already done like that across the app.

AdnanZiSAP commented on src/xcode/ENA/ENA/Source/Views/Common/HtmlTextView.swift:
```diff
@@ -47,7 +47,9 @@ class HtmlTextView: UITextView {
 
+		isUserInteractionEnabled = true
 		isEditable = false
+		dataDetectorTypes = [.link, .phoneNumber]
```
This text view is also used in Privacy screen in Onboarding. This change shouldn't break anything, but if you didn't I would still check that are links are still working there.

EisvogelSAP replied:
Jumping out of the onboarding flow was strongly discouraged during design discussions.

AdnanZiSAP replied:
Now it works, but tapping on an email there crashes the app. I will open a new discussion for that.

melloskitten replied:
I agree with @EisvogelSAP , I will inform the PO about this.

EisvogelSAP replied:
Has this been aligned with PO?

melloskitten replied:
@EisvogelSAP The PO said to follow the issue. In the issue, only the AppInformation screens are supposed to be interactive (so NOT the onboarding stuff). But this should be already handled by me moving this piece of code outside of the HTMLTextView.

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationDetailViewController.swift:
```diff
@@ -57,7 +57,7 @@ extension AppInformationDetailViewController {
 extension AppInformationDetailViewController: UITextViewDelegate {
 	func textView(_ textView: UITextView, shouldInteractWith url: URL, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
-		WebPageHelper.openSafari(withUrl: url, from: self)
+		UIApplication.shared.open(url, options: [:], completionHandler: nil)
```
Technically this can be removed altogether, as iOS handles links and phone number automatically if no delegate is provided, if I remember correctly. Please double check.

AdnanZiSAP replied:
I would double on this too.

melloskitten replied:
could be removed!

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Views/Common/HtmlTextView.swift:
```diff
@@ -47,7 +47,9 @@ class HtmlTextView: UITextView {
 			NSAttributedString.Key.foregroundColor: UIColor.enaColor(for: .textTint)
 		]
 
+		isUserInteractionEnabled = true
 		isEditable = false
+		dataDetectorTypes = [.link, .phoneNumber]
```
This PR does not take into account all occurrences of this component. So please double check if this generic application of the data detector types is applicable.

melloskitten replied:
moved away from this class, is specific to the AppInformationDetailViewController, which should have the clickable links 😄 the generic component is not touched.

melloskitten commented:
Hi! Thanks for reviewing. I have moved the entire functionality of being able to interact with the phone numbers/links/... away from the HTMLTextView class 👍

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationViewController+DynamicTableViewModel.swift:
```diff
@@ -64,10 +64,16 @@ private extension DynamicCell {
 	}
 
-	static func html(url: URL?) -> Self {
+	static func html(url: URL?, isInteractive: Bool = false) -> Self {
```
Why is there `isInteractive`? Is it ever used with `false` in this context?

melloskitten replied:
I wanted to leave the option to have html cells that are non-interactive, in order to be as little as invasive into the existing code base as possible. Is there any reason why this should not be done?

melloskitten replied:
Just got confirmed that this cell is used absolutely nowhere else. I will remove the parameter, as all .html cells are supposed to be interactive.

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationDetailViewController.swift:
```diff
@@ -54,10 +54,3 @@ extension AppInformationDetailViewController {
 	}
 }
-
-extension AppInformationDetailViewController: UITextViewDelegate {
-	func textView(_ textView: UITextView, shouldInteractWith url: URL, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
-		WebPageHelper.openSafari(withUrl: url, from: self)
-		return false
-	}
-}
```
So links within App Information will henceforth always open outside of the app context? Is this the desired effect and aligned with PO and/or design?

melloskitten replied:
I have not been able to get anything reliable out of the PO besides that the interactivity should only be in the AppInformation screens - due to this missing information, I will go with what @AdnanZiSAP suggested and make it consistent.

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationViewController+DynamicTableViewModel.swift:
```diff
@@ -218,7 +224,7 @@ extension AppInformationViewController {
 					text: AppStrings.AppInformation.privacyTitle,
 					accessibilityIdentifier: AccessibilityIdentifiers.AppInformation.privacyTitle),
-				.html(url: Bundle.main.url(forResource: "privacy-policy", withExtension: "html"))
+				.html(url: Bundle.main.url(forResource: "privacy-policy", withExtension: "html"), isInteractive: true)
```
See above

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Scenes/AppInformation/AppInformationViewController+DynamicTableViewModel.swift:
```diff
@@ -233,7 +239,7 @@ extension AppInformationViewController {
 					text: AppStrings.AppInformation.termsTitle,
 					accessibilityIdentifier: AccessibilityIdentifiers.AppInformation.termsTitle),
-				.html(url: Bundle.main.url(forResource: "usage", withExtension: "html"))
+				.html(url: Bundle.main.url(forResource: "usage", withExtension: "html"), isInteractive: true)
```
See above

EisvogelSAP commented on src/xcode/ENA/ENA/Source/Workers/WebPageHelper.swift:
```diff
@@ -37,7 +37,17 @@ enum WebPageHelper {
 	}
 
-	static func openSafari(withUrl url: URL, from viewController: UIViewController) {
+	static func open(withUrl url: URL, from viewController: UIViewController) {
+		switch url.scheme {
+		case "tel", "mailto":
+			UIApplication.shared.open(url, options: [:], completionHandler: nil)
+		default:
+			openLink(withUrl: url, from: viewController)
+		}
+
+	}
```
Semantically this is a little weird. It is a `WebPageHelper`. However it also handles phone and mail links. Would be cleaner to extract into a separate helper.

AdnanZiSAP replied:
Totally agree, this should be moved outside, either that or the helper should be renamed.

melloskitten replied:
I renamed the helper 👍

AdnanZiSAP commented:
SonarCloud wants you to add some unit tests for the new code you've introduced. :)

AdnanZiSAP commented on src/xcode/ENA/ENA/Source/Workers/WebPageHelper.swift:
```diff
@@ -37,7 +37,17 @@ enum WebPageHelper {
+
+	}
+
+	static func openLink(withUrl url: URL, from viewController: UIViewController) {
```
To me it looks like the complete code in this method is a duplicated code from ```showWebPage``` method. Ideally, this should be refactored so that only one method exists, when this code is already touched.

melloskitten replied:
hm, that's interesting. Since i just moved the already existing code into another method, i believe this duplication was already there before i touched anything at all. But i agree with you, this is an opportunity to remove duplicate code, so i will go ahead and do that! 💯 👍

melloskitten replied:
(I moved around the parameters as well, because they were inconsistent with openLink.)

AdnanZiSAP commented on src/xcode/ENA/ENA/Source/Scenes/Onboarding/OnboardingInfoViewController.swift:
```diff
@@ -372,7 +372,7 @@ final class OnboardingInfoViewController: UIViewController {
 extension OnboardingInfoViewController: UITextViewDelegate {
 	func textView(_ textView: UITextView, shouldInteractWith url: URL, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
-		WebPageHelper.openSafari(withUrl: url, from: self)
+		WebPageHelper.open(withUrl: url, from: self)
```
Could you please double check that this change has actually any effect in practice, i.e. that links are really opened in Onboarding?

On my device they're not, and I suspect it will be the same on all devices since ```dataDetectorTypes``` were removed from ```HtmlTextView```. 

Furthermore, based on concerns that Marc-Peter already raised, I would also suggest to align with PO on this, as apparently this was discouraged by PO/Design in the original implementation.

AdnanZiSAP replied:
Just checked the ticket. Based on the ticket, there's no mention of Onboarding at all. Hence I would suggest rather to don't touch Onboarding here at all, since concerns were raised, and basically no-one asked for changes there.

melloskitten replied:
I have changed it back so that the OnboardingViewController only uses the functionality for opening webpages like it was using it before (no switch statement, just plain old method).

melloskitten commented:
@AdnanZiSAP and @EisvogelSAP Thanks a bunch for your review. I have made sure that the opening of web links in the App Information is now consistent, as @AdnanZiSAP suggested. The Onboarding flow has the same logic of opening web links as before (I renamed the method, that's why it shows up as a change). I have also moved out the `isInteractive` property as @EisvogelSAP has confirmed that these cells are absolutely not used in any other part of the app 👍   Oh and I cleaned up the duplicate code that was already existing in the code base prior to my changes. Would be great if you took a look at it again! 

@AdnanZiSAP I don't see how to make any useful tests out of this PR, but I would greatly appreciate a hint if you already have something in mind how to test it. Cheers! 👍

AdnanZiSAP commented:
> @AdnanZiSAP and @EisvogelSAP Thanks a bunch for your review. I have made sure that the opening of web links in the App Information is now consistent, as @AdnanZiSAP suggested. The Onboarding flow has the same logic of opening web links as before (I renamed the method, that's why it shows up as a change). I have also moved out the `isInteractive` property as @EisvogelSAP has confirmed that these cells are absolutely not used in any other part of the app 👍 Oh and I cleaned up the duplicate code that was already existing in the code base prior to my changes. Would be great if you took a look at it again!
> 
> @AdnanZiSAP I don't see how to make any useful tests out of this PR, but I would greatly appreciate a hint if you already have something in mind how to test it. Cheers! 👍

Looks fine now, and seems to be working fine too. 👍

As for tests, I don't have any suggestion really, now I have it in my merge request, just as I merged develop into it and added one argument to a method. In general, I added one new property with like 5 lines of code. 🤔 I don't know why this was set up like it was really. Perhaps we should reconsider this.

AdnanZiSAP commented:
@melloskitten I will approve this as soon as you merge develop into this branch.

melloskitten commented:
@AdnanZiSAP 

> As for tests, I don't have any suggestion really, now I have it in my merge request, just as I merged develop into it and added one argument to a method. In general, I added one new property with like 5 lines of code. 🤔 I don't know why this was set up like it was really. Perhaps we should reconsider this.

I could not agree more with you. I do see the point of the quality gates, but they do not make sense for some of the changes that are currently done...
